src/passes/MemoryPacking.cpp

```diff
@@ -459,48 +461,53 @@ void MemoryPacking::getSegmentReferrers(Module* module,
         referrers[curr->segment].push_back(curr);
       }
       void doWalkFunction(Function* func) {
-        referrers.resize(getModule()->memory.segments.size());
         super::doWalkFunction(func);
       }
     } collector(referrers);
     collector.walkFunctionInModule(func, module);
   };
-  ModuleUtils::ParallelFunctionAnalysis<std::vector<Referrers>> analysis(
+  ModuleUtils::ParallelFunctionAnalysis<ReferrersMap> analysis(
     *module, collectReferrers);
-  referrers.resize(module->memory.segments.size());
-  for (auto& pair : analysis.map) {
-    std::vector<Referrers>& funcReferrers = pair.second;
-    for (size_t i = 0; i < funcReferrers.size(); ++i) {
+  for (auto& [_, funcReferrersMap] : analysis.map) {
+    for (auto& [i, segReferrers] : funcReferrersMap) {
       referrers[i].insert(
-        referrers[i].end(), funcReferrers[i].begin(), funcReferrers[i].end());
+        referrers[i].end(), segReferrers.begin(), segReferrers.end());
     }
   }
 }
 
 void MemoryPacking::dropUnusedSegments(std::vector<Memory::Segment>& segments,
-                                       std::vector<Referrers>& referrers) {
+                                       ReferrersMap& referrers) {
   std::vector<Memory::Segment> usedSegments;
-  std::vector<Referrers> usedReferrers;
+  ReferrersMap usedReferrers;
   // Remove segments that are never used
   // TODO: remove unused portions of partially used segments as well
+  size_t newSegIndex = 0;
   for (size_t i = 0; i < segments.size(); ++i) {
     bool used = false;
+    auto referrersIt = referrers.find(i);
+    bool hasReferrers = referrersIt != referrers.end();
     if (segments[i].isPassive) {
-      for (auto* referrer : referrers[i]) {
-        if (referrer->is<MemoryInit>()) {
-          used = true;
-          break;
+      if (hasReferrers) {
+        for (auto* referrer : referrersIt->second) {
+          if (referrer->is<MemoryInit>()) {
+            used = true;
+            break;
+          }
         }
       }
     } else {
+      // Active segment.
       used = true;
     }
     if (used) {
-      usedSegments.push_back(segments[i]);
-      usedReferrers.push_back(referrers[i]);
-    } else {
+      usedSegments.push_back(std::move(segments[i]));
```

Comment:
This modifies an input parameter which seems slightly surprising. Maybe worth mentioning in a comment on the definition?

Reply:
Oh, I see there is a `swap` below, so this always modified it...